##### ❗Προσοχή: Συμπληρώστε τα ονόματα των μελών της ομάδας σας

     Ονοματεπώνυμο 1ου μέλους: Τάκης Σταματόπουλος
             email 1ου μέλους: el20096@mail.ntua.gr
  
     Ονοματεπώνυμο 2ου μέλους: Κυριακή Καρατζούνη
             email 2ου μέλους: el20634@mail.ntua.gr


# Τεχνητή Νοημοσύνη: Εργαστηριακή Άσκηση 2
---

Ο στόχος της εργασίας είναι η κατασκευή ενός συστήματος προτάσεων (Recommendation System) για ταινίες. Οι προτάσεις αυτές θα πηγάζουν τόσο από τα χαρακτηριστικά της ταινίας όσο και από ορισμένες αξιολογήσεις του κάθε χρήστη.

Στα δεδομένα της άσκησης περιλαμβάνονται ένα αρχείο με το όνομα movies_metadata.csv το όποιο περιέχει τα χαρακτηριστικά κάθε ταινίας όπως θέμα, σκηνοθέτης ηθοποιοί, λέξεις κλειδιά κ.α. από το imdb καθώς και τα αρχεία ratings.csv τα όποια περιέχουν πραγματικές αξιολογήσεις χρηστών, χωρισμένες σε train και σε test.


# Εκφώνηση
Στην παρούσα εργασία σας ζητείται να μελετήσετε και να υλοποιήσετε τα παρακάτω:<br>

### Μέρος 1
Στο μέρος 1 και 2 θα εργαστείτε μόνο με το αρχείο movies_metadata.csv, ενώ στο μέρος τρία θα δουλέψετε και με τα αρχεία των αξιολογήσεων.
#### Ερώτημα 1α

Αρχικά θα πρέπει, αφού μελετήσετε τη δομή και τα χαρακτηριστικά του movies_metadata.csv, να κατασκευάσετε μια βάση γνώσης για την Prolog η όποια ουσιαστικά θα αποτελεί τον κόσμο με τον όποιο θα εργαστείτε στην συνέχεια. Τα κατηγορήματα που θα δημιουργηθούν θα σας βοηθήσουν και στην κατασκευή του recommender και θα είναι της μορφής:

```
director(Movie, Director).
genre(Movie, Genre).
```
#### Ερώτημα 1β

Αφού δημιουργήσετε τον κόσμο του προβλήματος, στη συνέχεια καλείστε να δημιουργήσετε, σε Prolog, απλούς κανόνες οι οποίοι θα βρίσκουν όλες τις ταινίες με:
1.	Κοινό θέμα (κάποιες λέξεις σχετικά με το genre κοινές)
2.	Αρκετά κοινό θέμα (κάποιες λιγότερες λέξεις σχετικά με το genre κοινές π.χ. 3)
3.	Σχετικά κοινό θέμα (λίγες λέξεις σχετικές με το genre κοινές π.χ. 1)
4.	Κοινός σκηνοθέτης
5.	Ακριβώς ίδια πλοκή (κάποιες λέξεις κλειδιά της πλοκής κοινές)
6.	Σχετικά ίδια πλοκή (κάποιες λιγότερες λέξεις κλειδιά κοινές)
7.	Ίδιους τους βασικούς ηθοποιούς (και τους 3)
8.	Αρκετά ίδιους βασικούς ηθοποιούς (ορισμένους βασικούς ηθοποιούς κοινούς π.χ. 2)
9.	Σχετικά ίδιους ηθοποιούς (π.χ. 1 από τους 3)
10.	Ίδια γλώσσα
11.	Είναι έγχρωμες ή ασπρόμαυρες
12. Κοινό studio παραγωγής
13. Κοινή χώρα παραγωγής
14. Ίδια δεκαετία

Αξίζει να σημειωθεί ότι στα παραπάνω μπορείτε να προσθέσετε περισσότερα ερωτήματα ή να αλλάξετε την κλιμακωσιμότητα των queries (πέρα από το ίδιο, αρκετά ίδιο, σχετικά ίδιο) που θα κατασκευάσετε(π.χ. μια κλίμακα από 1 στα 5 όπου αυτό είναι δυνατόν), μιας και αυτά στην συνέχεια θα χρησιμοποιηθούν για την λειτουργία του recommender. Έτσι μπορείτε να προσθέσετε queries με τα όποια μπορεί να παράγονται καλύτερες συστάσεις  (το αρχείο movie_metadata.csv περιέχει πολλές πληροφορίες ακόμα για κάθε ταινία όπως έτος κυκλοφορίας, βαθμολογία στο imdb, facebοok_likes κ.α.). Το συγκεκριμένο μέρος εργασίας είναι προπαρασκευαστικό οπότε όσο καλύτερα και πλουσιότερα τα ερωτήματα που θα φτιάξετε σε αυτό το μέρος τόσο καλύτερη θα είναι η απόδοσή των συστημάτων συστάσεων των επόμενων ερωτημάτων.

Περισσότερες πληροφορίες για το dataset μπορείτε να διαβάσετε σε αυτό το [Link](https://www.kaggle.com/georgefila/movies-metadata).
### Μέρος 2: Recommendation System
Στο σημείο αυτό καλείστε με βάση αυτά που κάνατε στο μέρος 1 να κατασκευάσετε queries τα όποια θα σας επιστρέφουν παρόμοιες (σε χαρακτηριστικά) ταινίες. Τα ερωτήματα αυτά θα είναι κλιμακούμενα, δηλαδή θα υπάρχουν ερωτήματα που επιστρέφουν αρκετά κοινές ταινίες αλλά και που επιστρέφουν λιγότερο και λιγότερο κοινές (σε μια κλίμακα π.χ. από 1 σε 5). Για παράδειγμα:

```
find_simmilar_movies_5("Pirates Of The Caribbean", M).
M = "Pirates Of The Caribbean: On Stranger Tides"
M = "The Chronicles Of Narnia"
M = "Prince Of Persia: The Sands Of Time"
...
```
Για παράδειγμα, το παραπάνω ερώτημα θα επιστρέφει αρκετά κοινές σε περιεχόμενο ταινίες με την ταινια "Pirates Of The Caribbean". Θα υπάρχουν και αντίστοιχα ερωτήματα που θα βρίσκουν λιγότερο όμοιες ταινίες. Ο δείκτης ομοιότητας των ταινιών είναι αυθαίρετος και μπορείτε να τον ορίσετε εσείς όπως θέλετε, αρκεί να υπάρχει κάποια λογική σύνδεση με τα δεδομένα που περιέχονται στο αρχείο movies_metadata.csv.

Συνεπώς η συνάρτηση που θα κάνει τις προτάσεις (recommendation) με είσοδο μια ταινία πρέπει να επιστρέφει-εκτυπώνει μια λίστα με τις προτεινόμενες ταινίες κατά φθίνουσα σειρά ομοιότητας.


###3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις - Αξιολογήσεις του χρήστη


Σε αυτό το σημείο θα εργαστείτε με τα αρχεία ratings τα όποια περιέχουν αξιολογήσεις (από 1 μέχρι 5) για τις παραπάνω ταινίες. Το προηγούμενο σύστημα συστάσεων προτείνει στον χρήστη ταινίες αποκλειστικά με βάση την ομοιότητά τους. Σε αυτό το σημείο θα γίνει μια αναβάθμιση του συστήματος έτσι ώστε να παράγονται καλύτερες  συστάσεις οι όποιες θα λαμβάνουν υπόψιν και τις προτιμήσεις του χρήστη, οι όποιες θα εξάγονται από τις αξιολογήσεις που έχει κάνει μέχρι στιγμής.

Η εκπαίδευση του recommender θα γίνεται ως εξής:

Για κάθε ταινία θα υπάρχει ένα score το όποιο αρχικά θε είναι ίσο με 0 και θα διαμορφώνεται από τις αξιολογήσεις κάθε user. Έτσι για έναν χρήστη με βάση τις αξιολογήσεις που υπάρχουν στο αρχείο train_ratings θα πρέπει:

1.	Για κάθε ταινία που έχει βαθμολογήσει να βρίσκονται οι κοινές ταινίες ανά κλίμακα και στο μέχρι τώρα σκορ κάθε παρόμοιας ταινίας θα προστίθεται ένα βάρος το όποιο θα μπορούσε να είναι το ποσοστό ομοιότητας της ταινίας (δηλαδή ένα βάρος για κάθε κλίμακα, αν δύο ταινίες μοιάζουν ενισχύουμε το βάρος που προσθέτουμε από το να μοιάζουν λιγότερο) επί τον βαθμό που έχει βάλει ο χρήστης για την αρχική ταινία (διαφορετικό είναι ο χρήστης να έχει βάλει 5/5 ή 1/5 σε μια ταινία από 3/5).

2. Στη συνέχεια, ανάλογα με το σκορ που έχει σχηματιστεί για κάθε ταινία, θα επιλέγεται αν αυτή θα μπορούσε να είναι προτεινόμενη για τον χρήστη ή όχι και θα μετράμε το πόσο καλά τα πήγε το σύστήμά μας με βάση ορισμένες μετρικές.

Η λογική πίσω από την παραπάνω διαδικασία είναι ότι παρόμοιες ταινίες θα έχουν ανάλογο βαθμό. Για παράδειγμα αν ένας χρήστης έχει αξιολογήσει αρκετές  ταινίες οι όποιες είναι sci-fiction με 5/5 τότε μια ταίνια sci-fiction την όποια δεν έχει δεί λογικά θα του αρέσει και θα έπρεπε να την προτείνουμε.

Μετά την εκπαίδευση του συστήματος σας καλείστε να δοκιμάσετε τον recommender που κατασκευάσατε στην πράξη. Για τον σκοπό αυτό θα φορτώσετε το αρχείο test_ratings.csv όπου περιέχονται οι αξιολογήσεις του ίδιου χρήστη για άλλες ταινίες. Το σύστημά σας πρέπει να προβλέπει αν μια ταινία θα πρέπει να προταθεί στον χρήστη. Μια ταίνια θεωρούμε οτι προτείνεται στον  χρήστη, αν έχει βαθμό μεγαλύτερο του 3 στο test_ratings. Συνεπώς για την επίβλεψη του συστήματός σας θα πρέπει για κάθε μια από τις ταινίες του αρχείου test_ratings.csv να προβλέψετε εάν θα αρέσει στο χρήστη ή όχι ώστε να την προτείνετε.

ΠΡΟΣΟΧΗ! Η "βαθμολογία" που θα υπολογίσετε για κάθε ταινία κατά το training δεν είναι απαραίτητα πρόβλεψη της βαθμολογίας που θα έβαζε ο χρήστης.

Στην συνέχεια, σε συνδιασμό με τις πραγματικές απαντήσεις του χρήστη θα αξιολογήσετε το σύστημά σας χρησιμοποιόντας τις μετρικές: precision, recall, f1 οι οποίες είναι οι πλέον γνωστές μετρικές και ευρέως χρησιμοποιούμενες τεχνικές για την επίβλεψη-μέτρηση απόδοσης ανάλογων συστημάτων.

1. Precision: Δείχνει πόσο ακριβές είναι το σύστημα. Υπολογίζει πόσα από τα στιγμιότυπα τα όποια προβλέψαμε ότι ανήκουν σε μια κλάση όντως ανήκουν σε αυτή. Η μετρική αυτή μας δίνει μια εικόνα σχετικά με τον αριθμό των ταινίων που προβλέψαμε ως προτεινόμενες ενώ δεν θα έπρεπε.

2. Recall: Υπολογίζει πόσα από τα στιγμιότυπα που ανήκουν σε μια κλάση (π.χ. προτεινόμενες ταινίες) προβλέφθηκαν σωστά.

3. F1: Είναι ένας μέσος μεταξύ των παραπάνω δυο μετρικών, έτσι ώστε να διατηρείται μια ισορροπία μεταξύ τους. Υπολογίζεται από την παρακάτω σχέση:

$$F_1=2\frac{Precision\times{Recall}}{Precision+Recall}$$

Οι παραπάνω συναρτήσεις παρέχονται από την βιβλιοθήκη scikit-learn.

Τέλος για την καλύτερη επίβλεψη του συστήματος σας μπορείτε να εκπαιδεύσετε  τον recommender σας με ένα υποσύνολο ταινίων από ελάχιστες, λίγες μέχρι και πολλές (π.χ. 3, 5, 10, 50, ...) για να μελετήσετε κατά πόσο σας βοηθούν οι επιπλέον αξιολογήσεις κάθε φορά (δηλαδή κατά πόσο βελτιώνονται οι παραπάνω μετρικές στο test set).

Έτσι π.χ. μπορείτε να εντοπίσετε περιπτώσεις όπως για παράδειγμα ότι με έναν recommender μπορεί να μην επιτυγχάνετε πολύ υψηλό σκορ όσο με άλλους, αλλά το βέλτιστο σκόρ σας επιτυγχάνεται πολύ γρηγόρα π.χ. με μόνο 10 ταινίες αντί 100. Έτσι για παράδειγμα αν ο αλγόριθμος τα πηγαίνει πολύ καλά για τρεις ταινίες και στη συνέχεια το σκορ βελτιώνεται ελάχιστα τότε αυτός ενδείκνυται για ένα σύστημα συστάσεων για νέους χρήστες όπου δεδομένου λίγων ταινιών ο αλγόριθμος είναι σε θέση να το να προτείνει καλές συστάσεις. Ενώ αν η καλύτερη απόδοση του αλγορίθμου σας είναι βέλτιστη με περισσότερες ταινίες π.χ. 50 τότε αυτός ο αλγόριθμος συστάσεων ενδείκνυται για παλιούς χρήστες με πολλές αξιολόγησεις.

---
## Κατασκευή Περιβάλλοντος Εργασίας
---
### Διάβασμα Αρχείων στο Colab (Μόνο για το Colab)
Αν η υλοποίηση γίνει στο google colab τότε μπορεί να χρησιμοποιηθεί το google drive ως file system. Για να γίνει Mount το google drive τρέχουμε τον παρακάτω κώδικα και κλικάρουμε στο link που θα μας εμφανιστεί.

```
from google.colab import drive
drive.mount('/content/drive')
import os
os.listdir('/gdrive/My Drive')
```
Έπειτα στην σελίδα που άνοιξε επιλέγουμε το mail μας και στο επόμενο παράθυρο που θα μας ανοίξει πατάμε Να επιτρέπεται. Στην συνέχεια αντιγράφουμε τον κωδικό που θα μας βγάλει και τον κάνουμε paste στο Input που έχει ανοίξει στο colab. Έτσι πλέον αν έχουμε ανεβάσει ένα αρχείο στο google drive μπορούμε να το βρούμε στην θέση:

```
movies_filename = '/gdrive/My Drive/' + movies_metadata.csv
```
Μπορούμε πλέον κανονικά να δουλέψουμε φτιάχνοντας φακέλους ή αρχεία και γενικότερα κάνοντας οτιδήποτε θα κάναμε αν ήμασταν τοπικά.

### Prolog μέσω Python

Το πακέτο που θα χρησιμοποιηθεί για την επικοινωνία Python και Prolog είναι το pyswip (https://pypi.org/project/pyswip/). Για να δουλέψει το Pyswip  χρειάζεται να υπάρχει το Swi-Prolog το όποιο αν δουλεύουμε τοπικά πρέπει να το εγκαταστήσουμε, ακολουθώντας αντίστοιχες οδηγίες στην σελίδα του εργαλείου. Για να γίνει του Swi-Prolog η εγκατάσταση στο Google Colab πρέπει να τρέξουμε τον παρακάτω κώδικα:

```
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog/stable
!sudo apt-get update
!sudo apt-get install swi-prolog
```
Σε κάποιο σημείο της εκτέλεσης εμφανίζεται ένα μήνυμα ότι πρέπει να πατήσουμε enter σε ένα input για να συνεχίσει η διαδικασία. Έπειτα από αυτό η εκτέλεση θα συνεχίσει χωρίς κάποιο πρόβλημα.

Τέλος πρέπει να εγκαταστήσουμε το pyswip (**όπου και να δουλεύουμε**) όπως παρακάτω:

```
!pip isntall pyswip
```


---
#Κώδικας για κατασκευή περιβάλλοντος εργασίας

##Μόνο για Google Colab

Κώδικας για να γίνει Mount to Google Drive

In [2]:
from google.colab import drive
drive.mount('/content/drive')

import os
path="" # με ευκόλο τρόπο μπορείτε να αλλάξετε το path που είναι αποθηκευμένα τα αρχεία σας

Mounted at /content/drive


In [3]:
%%capture
#install swi-prolog
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog
!sudo apt-get update
!sudo apt-get install swi-prolog
#install pyswip
!pip install pyswip


# **Μέρος 1: Μελέτη των Metadata, Δημιουργία κόσμου και των βασικών queries.**


In [4]:
import pandas as pd
from pyswip import Prolog

In [5]:
# Η βιβλιοθήκη pandas είναι χρήσιμη για την εργασία με τέτοια δεδομένα
import pandas as pd
path = '/content/drive/MyDrive/AI/'
# Διάβασμα του αρχείου 'movie_metadata.csv'
data = pd.read_csv(path + "movies_metadata.csv")
#Στο csv υπαρχούν κελία με nan τιμές
#Στις θέσεις αυτές βάζουμε 'UNK' πράγμα που  κάνουμε με την παρακάτω συνάρτηση
data.fillna("UNK", inplace=True)
# Preview the first 5 lines of the loaded data
data.head()

Unnamed: 0     budget                                    genres  \
0           0  237000000  Action|Adventure|Fantasy|Science Fiction   
1           1  300000000                  Adventure|Fantasy|Action   
2           2  245000000                    Action|Adventure|Crime   
3           3  250000000               Action|Crime|Drama|Thriller   
4           4  260000000          Action|Adventure|Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                       plot_keywords  language  \
0  culture clash|future|space war|space colony|so...   English   
1  ocean|drug abuse|exotic island|east india trad...   English   
2  spy|based on novel|secret agent|sequel|mi6|bri...  Français   
3  dc comics|crime fighter|terrorist|secret ident...   English   
4  based on novel|mars|medallion|space travel|pri...   English   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ...  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   
3  Following the death of District Attorney Harve...  112.312950  ...   
4  John Carter is a war-weary, former military ca...   43.926995  ...   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                movie_title vote_average  num_voted_users  \
0                                    Avatar          7.2            11800   
1  Pirates of the Caribbean: At World's End          6.9             4500   
2                                   Spectre          6.3             4466   
3                     The Dark Knight Rises          7.6             9106   
4                               John Carter          6.1             2124   

  title_year                   country      director_name     actor_1_name  \
0     2009.0  United States of America      James Cameron      Zoe Saldana   
1     2007.0  United States of America     Gore Verbinski    Orlando Bloom   
2     2015.0            United Kingdom         Sam Mendes  Christoph Waltz   
3     2012.0  United States of America  Christopher Nolan    Michael Caine   
4     2012.0  United States of America     Andrew Stanton     Lynn Collins   

       actor_2_name       actor_3_name  
0  Sigourney Weaver       Stephen Lang  
1   Keira Knightley  Stellan Skarsgård  
2       Léa Seydoux      Ralph Fiennes  
3       Gary Oldman      Anne Hathaway  
4   Samantha Morton       Willem Dafoe  

[5 rows x 27 columns]

Ο παραπάνω κώδικας εισάγει τις απαραίτητες βιβλιοθήκες, διαβάζει ένα αρχείο CSV που περιέχει μεταδεδομένα ταινιών σε ένα πλαίσιο δεδομένων pandas και αντικαθιστά τυχόν ελλείπουσες τιμές (NaN) με 'UNK'. Εμφανίζονται οι 5 πρώτες γραμμές του πλαισίου δεδομένων για να αποκτήσουμε μια γενική εικόνα της δομής των δεδομένων.

In [6]:
def clean_text(text):
  text = text.replace(u'\xa0', u'')
  text = text.replace(u"'", u'')
  return text

In [7]:
from sklearn.metrics._plot.confusion_matrix import ConfusionMatrixDisplay
import ast
#create World
#Ορίζουμε τον κόσμο μας
prolog = Prolog()

#Για κάθε row του πίνακα φτιάχνουμε τα κατηγορήματα που θέλουμε να αποθηκέυσουμε
#αρχικά σε μια λίστα με το όνομα literals
literals = []
movie_score = {}
for row in data.itertuples(index=True, name='Pandas'):
  movie_title = clean_text(getattr(row, 'movie_title'))
  director_name = clean_text(getattr(row, 'director_name'))
  plot_keywords= clean_text(getattr(row, 'plot_keywords'))
  title_year = clean_text(str(getattr(row, 'title_year')))
  country = clean_text(getattr(row, 'country'))


  if movie_title != "UNK":
    if director_name != "UNK":
      literals.append("director_name('"+ movie_title +"','" + director_name + "')")
    if title_year != "UNK":
      literals.append("title_year('"+ movie_title +"','" + title_year + "')")
      decade=title_year[:3]
      literals.append("title_decade('" + movie_title + "','" + decade + "')")
    if country != "UNK":
      literals.append("country('"+ movie_title +"','" + country + "')")


    color="colored"
    for plot_keyword in plot_keywords.split("|"):
      literals.append("keyword('"+ movie_title +"','"+ plot_keyword +"')")
      if plot_keyword == "black and white":
        color = "Black_and_White"
    literals.append("color('"+ movie_title +"','" + color + "')")


    for genre in getattr(row, 'genres').split("|"):
      literals.append("genre('"+ movie_title +"','"+ genre +"')")

    for actor in [getattr(row, 'actor_1_name'), getattr(row, 'actor_2_name'), getattr(row, 'actor_3_name')]:
      literals.append("actors('"+ movie_title +"','"+ actor.replace("'",'"') +"')")

    for language in getattr(row, 'language'):
      literals.append("language('"+ movie_title +"','"+ language +"')")

    movie_studios = ast.literal_eval(getattr(row, 'production_companies'))
    for studio in movie_studios:
      literals.append("studio('" + movie_title + "','" + clean_text(studio.get('name')) + "')")



#Η Prolog θέλει τα κατηγορήματά της με την σειρά
literals.sort()

for literal in literals:
  prolog.assertz(literal)
  #print (literal +'.')

#Επίσης μπορούμε να κάνουμε consult ένα έτοιμο αρχείο στον κόσμο όπως παρακάτω
#prolog.consult(path + "db.pl")

In [8]:
# prolog.assertz('(directed_by(X,Y) :- findall(M,director(M,X),Y))')
prolog.assertz('(similar_genre_4(X, Y) :-  genre(X,A), genre(Y,A), genre(X,B), genre(Y,B), genre(X,C), genre(Y,C), genre(X,D), genre(Y,D), A\\=B, A\\=C, A\\=D, B\\=C, B\\=D, C\\=D, X\\=Y)')
prolog.assertz('(similar_genre_3(X, Y) :-  genre(X,A), genre(X,B), genre(X,C), genre(Y,A), genre(Y,B), genre(Y,C), X\\=Y, A\\= B, A\\= C, B\\=C, not(similar_genre_4(X, Y)) )')
prolog.assertz('(similar_genre_2(X, Y) :-  genre(X,A), genre(X,B), genre(Y,A), genre(Y,B), X\\=Y, A\\= B, not(similar_genre_3(X, Y)) )')
prolog.assertz('(similar_genre_1(X, Y) :-  genre(X,A), genre(Y,A), X\\=Y, not(similar_genre_2(X, Y)) )')

prolog.assertz('(common_director(X, Y) :-  director_name(X,A), director_name(Y,A), X\=Y )')

prolog.assertz('(similar_keyword_4(X, Y) :-  keyword(X,A), keyword(Y,A), keyword(X,B), keyword(Y,B), keyword(X,C), keyword(Y,C), keyword(X,D), keyword(Y,D), A\\=B, A\\=C, A\\=D, B\\=C, B\\=D, C\\=D, X\\=Y)')
prolog.assertz('(similar_keyword_3(X, Y) :-  keyword(X,A), keyword(X,B), keyword(X,C), keyword(Y,A), keyword(Y,B), keyword(Y,C), X\\=Y, A\\= B, A\\= C, B\\=C, not(similar_keyword_4(X, Y)) )')
prolog.assertz('(similar_keyword_2(X, Y) :-  keyword(X,A), keyword(X,B), keyword(Y,A), keyword(Y,B), X\=Y, A\=B, not(similar_keyword_3(X,Y)) )')
prolog.assertz('(similar_keyword_1(X,Y ) :-  keyword(X,A), keyword(Y,A), X\=Y, not(similar_keyword_2(X,Y)) )')

prolog.assertz('(common_actors_3(X, Y) :-  actors(X,A), actors(X,B), actors(X,C), actors(Y,A), actors(Y,B), actors(Y,C), X\\=Y, A \\= B, A \\= C, B \\= C)')
prolog.assertz('(common_actors_2(X, Y) :-  actors(X,A), actors(X,B), actors(Y,A), actors(Y,B), X\\=Y, A \\= B, not(common_actors_3(X, Y)) )')
prolog.assertz('(common_actors_1(X, Y) :-  actors(X,A), actors(Y,A), X\\=Y, not(common_actors_2(X, Y)) )')

prolog.assertz('(common_language(X, Y) :- language(X,D), language(Y,D), X \= Y)')

prolog.assertz('(same_color(X, Y) :- color(X, C), color(Y, C), X \\= Y )')

prolog.assertz('(same_production_studio(X, Y) :- studio(X,A), studio(Y,A), X \= Y )')

prolog.assertz('(same_country(X, Y) :- country(X, C), country(Y, C), X \\= Y )')

prolog.assertz('(same_decade(X, Y) :- title_decade(X, A), title_decade(Y, A), X\\=Y )')


Αυτή η ενότητα ορίζει μια συνάρτηση για τον καθαρισμό δεδομένων κειμένου, αρχικοποιεί μια παρουσία Prolog και επεξεργάζεται τις γραμμές του dataframe για να δημιουργήσει γεγονότα Prolog (κυριολεκτικά). Στη συνέχεια, ταξινομεί αυτά τα γεγονότα και τα δηλώνει στη βάση γνώσεων Prolog. Επιπλέον, ορίζονται αρκετοί κανόνες Prolog για τον εντοπισμό ομοιοτήτων μεταξύ ταινιών βάσει ειδών, λέξεων-κλειδιών, σκηνοθετών, ηθοποιών, γλωσσών, χρωμάτων, στούντιο παραγωγής, χωρών και δεκαετιών.

# **2ο Μέρος: Recommendation System με βάση μόνο τα χαρακτηριστικά των ταινιών.**

Στο σημείο αυτό με βάση τους κανόνες που κατασκευάστηκαν στο Μέρος 1 θα κατασκευαστούν κατηγορήματα για την ομοιότητα ταινιών. Παρακάτω δίνεται ένα μικρό παράδειγμα ενός κανόνα και πώς αυτός θα μπορούσε να γραφτεί μέσω του Pyswip. Επίσης όπως αναφέρεται και σε σχόλιο παραπάνω θα μπορεί να γραφτεί και μια βάση δεδομένων με τους κανόνες και να γίνει απευθείας consult.

Στο παρακάτω παράδειγμα το 5 και το 4 εκφράζουν την ομοιότητα των ταινιών π.χ. οι ταινίες που παράγονται μέσω του find_similar_5 είναι πιο όμοιες από αυτές που παράγονται μέσω του find_similar_4.


In [9]:
prolog.assertz('(find_similar_movies_7(X, Y) :- similar_genre_4(X, Y), similar_keyword_4(X, Y), common_language(X, Y) )') # Extremely Similar Movies
prolog.assertz('(find_similar_movies_6(X, Y) :- similar_genre_3(X, Y), similar_keyword_3(X, Y), common_language(X, Y) )') # Very Similar Movies  \
prolog.assertz('(find_similar_movies_5(X, Y) :- similar_genre_2(X, Y), similar_keyword_3(X, Y), common_language(X, Y) )') # Quite Similar Movies  > Same NO Keywords,
prolog.assertz('(find_similar_movies_4(X, Y) :- similar_genre_1(X, Y), similar_keyword_3(X, Y), common_language(X, Y) )') # Quite Similar Movies /    Different NO Genres
prolog.assertz('(find_similar_movies_3(X, Y) :- similar_genre_2(X, Y), similar_keyword_2(X, Y), common_language(X, Y) )') # Similar
prolog.assertz('(find_similar_movies_2(X, Y) :- similar_genre_1(X, Y), similar_keyword_1(X, Y), common_language(X, Y) )') # Less Similar
prolog.assertz('(find_similar_movies_1(X, Y) :- similar_genre_1(X, Y), common_language(X, Y) )') # Only 1 Genre and Language in Common

In [10]:
def similarity_recommender(movie,similarity): # Recommend Movie based on Similarity Grade 1-7
    s = set()
    q = prolog.query("find_similar_movies_"+str(similarity)+"('" + movie +"',M)")
    for soln in q:
        m = soln['M']
        if m not in s:
            s.add(soln['M'])
    q.close()
    answers = s
    return answers,similarity

Η συνάρτηση similarity_recommender λαμβάνει το όνομα μιας ταινίας και το επίπεδο ομοιότητας και επιστρέφει ταινίες που είναι παρόμοιες με βάση το genre, τις λέξεις-κλειδιά και τη γλώσσα.

In [11]:
def keyword_based_recommender(movie,similarity): # Recommend Movie Based on Keyword Similarity Grade 1-4
  s = set()
  q = prolog.query("similar_keyword_" + str(similarity) + "('" + movie + "',M)")
  for soln in q:
    m = soln['M']
    if m not in s:
      s.add(soln['M'])
  q.close()
  answers = s
  return answers,similarity

Η συνάρτηση keyword_based_recommender λειτουργεί με παρόμοιο τρόπο, αλλά βασίζεται μόνο στις λέξεις-κλειδιά.

In [12]:
def choose_better_recommendation(movie,similarity,KBSimilarity): # Choose Recommended Movie based on Similarity or Keyword
  similar_answers, sim1 = similarity_recommender(movie,similarity)
  KB_answers, sim2 = keyword_based_recommender(movie,KBSimilarity)
  if similar_answers:
    return similar_answers # Return similar movies based on find_similar_movies
  elif KB_answers:
    return KB_answers # Return movies based on similar number of keywords
  else:
    if sim1 == 1:
      if sim2 > 1:
        return choose_better_recommendation(movie,sim1,sim2-1) # If find_simila_movies == 0, check based only on keywords
      else:
        return []
    return choose_better_recommendation(movie,similarity-1,KBSimilarity-1) # Check for next lower grade similarity

Η συνάρτηση choose_better_recommendation συνδυάζει τις δύο προηγούμενες συναρτήσεις για να επιλέξει την καλύτερη δυνατή σύσταση, είτε με βάση το genre είτε με βάση τις λέξεις-κλειδιά.

In [13]:
def advanced_recommender(movie): # Advanced Recommender for all grades
  sim = 4 # starting number of matching keywords
  max_movies = 3 # max number of movies shown
  for x in range(7):
    print("For the movie '" + movie + "', and similarity " + str(x+1) + ", I suggest: " + str(list(choose_better_recommendation(movie,x+1,sim))[:max_movies]))

Η συνάρτηση advanced_recommender χρησιμοποιεί διαφορετικά επίπεδα ομοιότητας για να προτείνει ταινίες. Ορίζει αρχικά ότι οι λέξεις-κλειδιά πρέπει να ταιριάζουν σε 4 περιπτώσεις και εμφανίζει μέχρι 3 ταινίες για κάθε επίπεδο ομοιότητας από το 1 έως το 7.

In [14]:
def simple_recommender(movie):
    s = set()
    q = prolog.query("find_similar_movies_1('" + movie +"',M)")
    for soln in q:
        m = soln['M']
        if m not in s:
            s.add(soln['M'])
    q.close()
    answers = s
    return answers

Η συνάρτηση simple_recommender βρίσκει ταινίες που έχουν τουλάχιστον ένα κοινό genre και την ίδια γλώσσα με την ταινία εισόδου.

In [15]:
#list(simple_recommender('Avatar'))[:5]
advanced_recommender('Avatar')

For the movie 'Avatar', and similarity 1, I suggest: ['The Wild Bunch', 'The Sorcerers Apprentice', 'Blood Diamond']
For the movie 'Avatar', and similarity 2, I suggest: ['Transformers: Age of Extinction', 'The Wild Bunch', 'Supernova']
For the movie 'Avatar', and similarity 3, I suggest: ['Jimmy Neutron: Boy Genius', 'Alien', 'Jupiter Ascending']
For the movie 'Avatar', and similarity 4, I suggest: ['Galaxina', 'Pirates of the Caribbean: On Stranger Tides', 'Predators']
For the movie 'Avatar', and similarity 5, I suggest: ['Aliens', 'Alien³', 'The Fifth Element']
For the movie 'Avatar', and similarity 6, I suggest: ['Pirates of the Caribbean: On Stranger Tides', 'Predators', 'John Carter']
For the movie 'Avatar', and similarity 7, I suggest: ['Star Trek Into Darkness']


# **3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις-Αξιολογήσεις του χρήστη-Εκπαίδευση και Πρόβλεψη**


Αρχικά μελετάμε τις αξιολογήσεις κάθε χρήστη για να καταλάβουμε την δομή και τις πληροφορίες κάθε αρχείου.




In [16]:
from tqdm.notebook import tqdm
from sklearn.metrics import precision_score, recall_score, f1_score
import numpy as np
import random

rating_weights = {0: -1, 1: -0.5, 2:0, 3:0, 4:0.5, 5:1}
score_weights = {i:i + 1 for i in range(1)} # ανάλογα με τα επίπεδα ομοιότητας που έχουν οριστεί στην simple_recommender

def train_recommender(ratings, rating_weights, score_weights, number_of_movies = 10):
    """
    Στην συνάρτηση αυτή μπορούμε να ορίζουμε ποιο υποσύνολο των αξιολογήσεων θα χρησιμοποιήσουμε για το train μαζί με τα βάρη ομοιότητας και σκορ
    Σε συνδυασμό με τον αριθμό των ταινιών που θέλουμε να χρησιμοποιήσουμε σαν σύνολο δεδομένων π.χ. 10 από τις 100 ή 3 από τις 100 κ.ο.κ
    Αν θέλουμε να χρησιμοποιήσουμε όλες τις ταινίες σαν training set τότε ορίζουμε το number_of_movies = - 1
    """

    if number_of_movies > len(ratings):
        number_of_movies = len(ratings)


    if number_of_movies != -1:
        indexes = random.sample(range(len(ratings)), number_of_movies)
        ratings = ratings.iloc[indexes]

    movie_score = {}
    for row in tqdm(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        similar_movies = simple_recommender(movie)

        for similar_movie in similar_movies:
            if similar_movie not in movie_score:
                movie_score[similar_movie] = rating_weights[int(rating)] * score_weights[0]
            else:
                movie_score[similar_movie] += rating_weights[int(rating)] * score_weights[0] # το weight θα το ορίσετε ανα επίπεδο ομοιότητας οι πολύ όμοιες ταινίες θα έχουν μεγαλύτερο βάρος
    return movie_score


# Αυτό είναι ένα παράδειγμα για το πως θα μπορούσε να υλοποιήθει η predict.
# Έχουμε ορίσει ότι μια ταινία θα έπρεπε να είναι προτεινόμενη αν είχε σκορ > 0.
def predict_example(ratings, movie_score):
    real, pred = [], []
    for i, row in enumerate(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        if movie in movie_score: #αν έχουμε σχηματίσει βαθμολογία για την ταινία αυτή
            pred.append(int(movie_score[movie] > 0)) #heuristic για το αν μια ταινία είναι προτεινόμενη
            real.append(int(rating > 3))# έτσι ορίζουμε ότι μια ταινία θα έπρεπε να είναι προτεινόμενη
            #η συνθήκη αυτή δεν μπορεί να αλλάξει
        else: #δεν μπορούμε να προτείνουμε κάτι για το όποιο δεν έχουμε σχηματίσει εικόνα
            pred.append(0)
            real.append(int(rating > 3))

    return real, pred


def get_metrics(real, pred):
    metrics = {}
    metrics["precision"] = precision_score(real, pred)
    metrics["recall"] = recall_score(real, pred)
    metrics["f1"] = f1_score(real, pred)
    return metrics

Η παραπάνω εκπαιδεύουν, τεστάρουν και μετρούν την απόδοσή του συστήματος στάσεων μας. Για την εκπαίδευση του συστήματος μπορούμε κάθε φορά να χρησιμοποιήσουμε ένα τυχαίο υποσύνολο του training set. Όμως είναι πιθανό το υποσύνολο των ταινιών αυτόν να επηρεάζει τα αποτελέσματα στο training set. Για παράδειγμα από 3 ταινίες στις 10 μπορούν για ένα πείραμα τα αποτελέσματα μας αν είναι ίδια και αυτό να μην οφείλεται στο γεγονός ότι ο ταξινομητής μας τα πηγαίνει καλά στις 3 ταινίες αλλά στο γεγονός ότι οι υπόλοιπες 7 είναι τέτοιες ταινίες που δεν μας βοηθούν καθόλου και αν είχαμε επιλέξει διαφορετικές 3 ταινίες να τα πηγαίναμε χάλια. Οπότε προτείνουμε να τρέξετε κάθε φορά έναν αριθμό πειραμάτων για κάθε υποσύνολο ταινιών π.χ. 10 πειράματα με 3 ταινίες, 10 πειράματα για 20 ταινίες κ.ο.κ. και να κρατήσετε σαν τελικό σκορ το μέσο όρο όλων των πειραμάτων. Έτσι τα αποτελέσματα των αγγλικών σας θα είναι πιο αντικειμενικά.

In [17]:
train_ratings = pd.read_csv(path + "train_ratings.csv")
test_ratings = pd.read_csv(path + "test_ratings.csv")

In [18]:
train_ratings.head()


Unnamed: 0  userId  movieId            movie_title  rating   timestamp
0       78477     547      235            Stand by Me     4.0   974780707
1       78747     547     1251  Letters from Iwo Jima     5.0   974809897
2       78808     547     1391      Y Tu Mamá También     0.5  1076809696
3       78584     547      824          Moulin Rouge!     2.0  1076968814
4       80322     547    51540        Horrible Bosses     4.0  1192372083

In [19]:
test_ratings.head()


Unnamed: 0  userId  movieId         movie_title  rating   timestamp
0       78436     547       11           Star Wars     3.0  1418149949
1       78437     547       14     American Beauty     3.5  1053171195
2       78438     547       16  Dancer in the Dark     4.0   986694981
3       78441     547       25             Jarhead     4.5  1053087861
4       78452     547      111            Scarface     5.0   974809870

In [20]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 10)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5233948357811995
recall: 0.7486111111111111
f1: 0.6124466533832874


In [21]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 30)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5245567499256516
recall: 0.861111111111111
f1: 0.6505307777735461


Αναλύουμε τα αποτελέσματα για διαφορετικά μεγέθη υποσυνόλων:

    Με 10 ταινίες:
        Precision: 0.523
        Recall: 0.749
        F1: 0.612

    Με 30 ταινίες:
        Precision: 0.525
        Recall: 0.861
        F1: 0.651

Συμπεράσματα

Τα αποτελέσματα δείχνουν ότι καθώς αυξάνουμε τον αριθμό των ταινιών που χρησιμοποιούνται για την εκπαίδευση, οι μετρικές ανάκλησης και F1 βελτιώνονται. Ωστόσο, η ακρίβεια παραμένει σχεδόν σταθερή. Αυτό υποδηλώνει ότι η προσθήκη περισσότερων ταινιών βοηθά στην καλύτερη αναγνώριση των προτεινόμενων ταινιών, αλλά η ακρίβεια των προτάσεων δεν βελτιώνεται σημαντικά.

MΕΡΟΣ 3ο:

In [22]:
def simple_recommender(movie):
  answers = set()
  rec_length = [0,0,0,0,0,0,0]
  rec = choose_better_recommendation(movie,1,4) # Grade 1, Weight 1
  answers = rec
  rec_length[0] = len(answers)
  for i in range(7):
    if i < 1:
      continue
    rec = choose_better_recommendation(movie,i+1,4) # Grade 2, Weight 2 -> Grade 3, Weight 3 -> ... -> Grade 7, Weight 7
    for newMovie in rec:
      if newMovie not in answers:
        answers.add(newMovie)
    rec_length[i] = len(answers) - rec_length[i-1]
  return answers,rec_length

In [23]:
from tqdm.notebook import tqdm
from sklearn.metrics import precision_score, recall_score, f1_score
import numpy as np
import random

rating_weights = {0: -1, 1: -0.5, 2:0, 3:0, 4:0.5, 5:1}
score_weights = {i:i + 1 for i in range(7)} # ανάλογα με τα επίπεδα ομοιότητας που έχουν οριστεί στην simple_recommender

def train_recommender(ratings, rating_weights, score_weights, number_of_movies = 10):
    """
    Στην συνάρτηση αυτή μπορούμε να ορίζουμε ποιο υποσύνολο των αξιολογήσεων θα χρησιμοποιήσουμε για το train μαζί με τα βάρη ομοιότητας και σκορ
    Σε συνδυασμό με τον αριθμό των ταινιών που θέλουμε να χρησιμοποιήσουμε σαν σύνολο δεδομένων π.χ. 10 από τις 100 ή 3 από τις 100 κ.ο.κ
    Αν θέλουμε να χρησιμοποιήσουμε όλες τις ταινίες σαν training set τότε ορίζουμε το number_of_movies = - 1
    """

    if number_of_movies > len(ratings):
        number_of_movies = len(ratings)


    if number_of_movies != -1:
        indexes = random.sample(range(len(ratings)), number_of_movies)
        ratings = ratings.iloc[indexes]

    movie_score = {}
    for row in tqdm(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        similar_movies, idx_list = simple_recommender(movie)
        idx = 0
        counter = 0

        for similar_movie in similar_movies:
            if counter < idx_list[0]:
              idx = 0
            elif counter < idx_list[1]:
              idx = 1
            elif counter < idx_list[2]:
              idx = 2
            elif counter < idx_list[3]:
              idx = 3
            elif counter < idx_list[4]:
              idx = 4
            elif counter < idx_list[5]:
              idx = 5
            else:
              idx = 6
            if similar_movie not in movie_score:
                movie_score[similar_movie] = rating_weights[int(rating)] * score_weights[idx]
            else:
                movie_score[similar_movie] += rating_weights[int(rating)] * score_weights[idx] # το weight θα το ορίσετε ανα επίπεδο ομοιότητας οι πολύ όμοιες ταινίες θα έχουν μεγαλύτερο βάρος
            counter += 1
    return movie_score


# Αυτό είναι ένα παράδειγμα για το πως θα μπορούσε να υλοποιήθει η predict.
# Έχουμε ορίσει ότι μια ταινία θα έπρεπε να είναι προτεινόμενη αν είχε σκορ > 0.
def predict_example(ratings, movie_score):
    real, pred = [], []
    for i, row in enumerate(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        if movie in movie_score: #αν έχουμε σχηματίσει βαθμολογία για την ταινία αυτή
            pred.append(int(movie_score[movie] > 0)) #heuristic για το αν μια ταινία είναι προτεινόμενη
            real.append(int(rating > 3))# έτσι ορίζουμε ότι μια ταινία θα έπρεπε να είναι προτεινόμενη
            #η συνθήκη αυτή δεν μπορεί να αλλάξει
        else: #δεν μπορούμε να προτείνουμε κάτι για το όποιο δεν έχουμε σχηματίσει εικόνα
            pred.append(0)
            real.append(int(rating > 3))

    return real, pred


def get_metrics(real, pred):
    metrics = {}
    metrics["precision"] = precision_score(real, pred)
    metrics["recall"] = recall_score(real, pred)
    metrics["f1"] = f1_score(real, pred)
    return metrics

In [24]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 10)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5122433502700436
recall: 0.7291666666666667
f1: 0.5901346458777426


In [25]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 30)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5153298010162957
recall: 0.826388888888889
f1: 0.6317071652240471


Αποτελέσματα και Συμπεράσματα
Με 10 ταινίες:

    Precision: 0.512
    Recall: 0.729
    F1: 0.590

Με 30 ταινίες:

    Precision: 0.515
    Recall: 0.826
    F1: 0.632

Τα αποτελέσματα δείχνουν ότι με την αύξηση του αριθμού των ταινιών που χρησιμοποιούνται για την εκπαίδευση, οι μετρικές recall και F1 βελτιώνονται, ενώ η precision παραμένει σχεδόν σταθερή. Αυτό υποδηλώνει ότι η προσθήκη περισσότερων ταινιών βοηθά στην καλύτερη αναγνώριση των προτεινόμενων ταινιών, αλλά η ακρίβεια των προτάσεων δεν βελτιώνεται σημαντικά.